includes/class-wc-download-handler.php

```diff
@@ -33,136 +33,111 @@ public static function init() {
 	 * Check if we need to download a file and check validity.
 	 */
 	public static function download_product() {
-		$product_id    = absint( $_GET['download_file'] );
-		$_product      = wc_get_product( $product_id );
-		$download_data = self::get_download_data( array(
-			'product_id'  => $product_id,
-			'order_key'   => wc_clean( $_GET['order'] ),
-			'email'       => sanitize_email( str_replace( ' ', '+', $_GET['email'] ) ),
-			'download_id' => wc_clean( isset( $_GET['key'] ) ? preg_replace( '/\s+/', ' ', $_GET['key'] ) : '' ),
-		) );
+		$product_id   = absint( $_GET['download_file'] );
+		$product      = wc_get_product( $product_id );
+		$data_store   = WC_Data_Store::load( 'customer-download' );
 
-		if ( $_product && $download_data ) {
-			self::check_current_user_can_download( $download_data );
-
-			do_action( 'woocommerce_download_product', $download_data->user_email, $download_data->order_key, $download_data->product_id, $download_data->user_id, $download_data->download_id, $download_data->order_id );
-
-			self::count_download( $download_data );
-			self::download( $_product->get_file_download_path( $download_data->download_id ), $download_data->product_id );
-		} else {
+		if ( ! $product || ! isset( $_GET['key'], $_GET['order'] ) ) {
 			self::download_error( __( 'Invalid download link.', 'woocommerce' ) );
 		}
-	}
-
-	/**
-	 * Get a download from the database.
-	 *
-	 * @param  array  $args Contains email, order key, product id and download id
-	 * @return object
-	 * @access private
-	 */
-	private static function get_download_data( $args = array() ) {
-		global $wpdb;
 
-		$query = "SELECT * FROM " . $wpdb->prefix . "woocommerce_downloadable_product_permissions ";
-		$query .= "WHERE user_email = %s ";
-		$query .= "AND order_key = %s ";
-		$query .= "AND product_id = %s ";
+		$download_ids = $data_store->get_downloads( array(
+			'user_email'  => sanitize_email( str_replace( ' ', '+', $_GET['email'] ) ),
+			'order_key'   => wc_clean( $_GET['order'] ),
+			'product_id'  => $product_id,
+			'download_id' => wc_clean( preg_replace( '/\s+/', ' ', $_GET['key'] ) ),
```

Comment:
Some mail clients insert line breaks within download links. This is a workaround.